src/DefaultBuilder/src/Configuration.cs

```diff
@@ -19,10 +18,11 @@ namespace Microsoft.AspNetCore.Builder
     public sealed class Configuration : IConfigurationRoot, IConfigurationBuilder, IDisposable
     {
         private readonly ConfigurationSources _sources;
-        private ConfigurationRoot _configurationRoot;
 
+        private readonly object _providerLock = new();
```

Comment:
I imagine this lock might be one of the more unexpected/contentious part of this change.

I think we need to lock over all access to  `_providers` and `_changeTokenRegistrations` but not `_sources`.

My argument is that nothing should be modifying the IConfigurationBuilder (which is only exposed via explicit interface implementation FWIW) in parallel since that would break with a normal ConfigurationBuilder. However, avoiding reading from the IConfiguration while sources are being added might be harder.

For instance, I would not be surprised if something stored the IConfiguration from the WebHostContext (which [very well could be this Configuration type](https://github.com/dotnet/aspnetcore/blob/b9d7c3298b7e47f4b8585504a684c1e7d1b87478/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs#L31-L35) and then attempt to read it later on a background thread while the app is still adding config sources.

Does anyone think I am being too paranoid about `_providers`? Not paranoid enough about `_sources`?

Reply:
Just to make sure I understand what you are saying, we are relying on convention that no one should expect things to behave properly if they modify sources, while we explicitly guard against providers being modified?  What's the behavior today, when does the new source become 'live' or visible to everyone?  Or is this just not something we need to worry about since we only expose the source APIs for hosting?

Reply:
> What's the behavior today, when does the new source become 'live' or visible to everyone?

Today you have to call IConfigurationBuilder.Build() to get the IConfiguration object to read from, so once you get an IConfiguration instance, the sources and providers are static.

This new Configuration type implements both IConfigurationBuilder and IConfiguration at the same time, and the IConfiguration updates without any explicit call to something like Build() or NotifySourcesChanged().

At least theoretically, something could be reading from the Configuration while the sources/providers are being mutated. We could try to say that's not allowed, but with all the layers, this might be a hard requirement to expect people to follow.

Reply:
Right, that was the idea behind sources/providers to make it clear that its immutable, sorry I wasn't clear, when I meant 'today', I was asking what happens if you add a source while someone is reading it in the current code, will the readers start seeing updated (and possibly inconsistent) config values immediately?  I recall there were bugs/weirdness in the past, where readers would get empty config due to errors when reloading config

Reply:
This is one of the issues we found where the config changed during binding: https://github.com/dotnet/extensions/pull/1202

Reply:
I think I understand what you're getting out now @HaoK. Today, the `IConfigurationRoot` is backed by a new `ConfigurationRoot` instance each time it's updated, so config should be consistent for any given access, though it does look like there's likely a [race with Dispose](https://github.com/dotnet/aspnetcore/blob/4cdd594dad20aef83ecdff5c72a6121da1a83180/src/DefaultBuilder/src/Configuration.cs#L74).

Of course, if multiple keys are accessed independently concurrently with a config source being added, that might lead to inconsistent data being read unless the ReloadToken is used to observe changes.